[SPARK-33509][SQL] List partition by names from a V2 table which supports partition management

MaxGekk opened the pull request:
### What changes were proposed in this pull request?
1. Add new method `listPartitionByNames` to the `SupportsPartitionManagement` interface. It allows to list partitions by partition names and their values.
2. Implement new method in `InMemoryPartitionTable` which is used in DSv2 tests. 

### Why are the changes needed?
Currently, the `SupportsPartitionManagement` interface exposes only `listPartitionIdentifiers` which allows to list partitions by partition values. And it requires to specify all values for partition schema fields in the prefix. This restriction does not allow to list partitions by some of partition names (not all of them).

For example, the table `tableA` is partitioned by two column `year` and `month`
```
CREATE TABLE tableA (price int, year int, month int)
USING _
partitioned by (year, month)
```
and has the following partitions:
```
PARTITION(year = 2015, month = 1)
PARTITION(year = 2015, month = 2)
PARTITION(year = 2016, month = 2)
PARTITION(year = 2016, month = 3)
```
If we want to list all partitions with `month = 2`, we have to specify `year` for **listPartitionIdentifiers()** which not always possible as we don't know all `year` values in advance. New method **listPartitionByNames()** allows to specify partition values only for `month`, and get two partitions:
```
PARTITION(year = 2015, month = 2)
PARTITION(year = 2016, month = 2)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By running the affected test suite `SupportsPartitionManagementSuite`.

MaxGekk committed:
Add listPartitionByNames() to the SupportsPartitionManagement interface

MaxGekk committed:
First implementation of listPartitionByNames()

MaxGekk committed:
Add a test

MaxGekk committed:
Gets all partitions

MaxGekk committed:
Add asserts

MaxGekk committed:
Nothing matches to parameters

SparkQA commented:
**[Test build #131472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131472/testReport)** for PR 30452 at commit [`3f20ee8`](https://github.com/apache/spark/commit/3f20ee8289107f10a65165e4a6dbb69250039efa).

MaxGekk commented:
@cloud-fan Could you take a look at this PR, please.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36079/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/36079/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36079/
Test PASSed.

SparkQA commented:
**[Test build #131472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131472/testReport)** for PR 30452 at commit [`3f20ee8`](https://github.com/apache/spark/commit/3f20ee8289107f10a65165e4a6dbb69250039efa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131472/
Test PASSed.

HyukjinKwon commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsPartitionManagement.java:
```diff
@@ -106,10 +106,19 @@ void replacePartitionMetadata(
+     * @param ident a partition identifier values.
+     * @return an array of Identifiers for the partitions
+     */
+    InternalRow[] listPartitionByNames(String[] names, InternalRow ident);
```
Is a user supposed to access to `Table` directly? Looks a bit odd that `listPartitionByNames` interface is added but not used in the Spark internal side.

cloud-fan replied:
Yea we should remove the old API as it's not released yet.

MaxGekk replied:
The method `listPartitionIdentifiers()` is used only in `partitionExists()` and in tests. Let me remove it and replace its usage by `listPartitionByNames()`.

How about renaming `listPartitionByNames()` to just `listPartitions()`?

MaxGekk replied:
> Looks a bit odd that listPartitionByNames interface is added but not used in the Spark internal side.

@HyukjinKwon `listPartitionByNames()` will be used by V2 commands like `SHOW PARTITIONS` and `SHOW TABLE EXTENDED`.

MaxGekk replied:
> Yea we should remove the old API as it's not released yet.

Frankly speaking, I would remove `listPartitionIdentifiers()` separately as this requires unrelated changes to list partition by names.

cloud-fan replied:
I think the final API name should still be `listPartitionIdentifiers`

cloud-fan replied:
We can remove the old one in your next PR.

stczwd replied:
> I think the final API name should still be `listPartitionIdentifiers`

Agree with this.

cloud-fan commented:
The partition can be a transform like `year(ts_col)`, shall we just partition index in the API instead?

MaxGekk commented:
As the changes are related to `listPartitionIdentifiers()` added by https://github.com/apache/spark/pull/28617. @stczwd @rdblue @RussellSpitzer @emkornfield @dongjoon-hyun May I ask you to review this PR.

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/connector/InMemoryPartitionTable.scala:
```diff
@@ -96,4 +97,25 @@ class InMemoryPartitionTable(
+    assert(names.forall(fieldName => schema.fieldNames.contains(fieldName)),
+      s"Some partition names ${names.mkString("[", ", ", "]")} don't belong to " +
+      s"the partition schema '${schema.sql}'.")
+    val indexes = names.map(schema.fieldIndex)
+    val dataTypes = names.map(schema(_).dataType)
```
The names should be normalized after https://github.com/apache/spark/pull/30454, so, we shouldn't care of case sensitivity here.

MaxGekk commented:
> ... shall we just partition index in the API instead?

@cloud-fan Sure. I see at least two variants of the implementation:
- Pass an array of indexes, or
- a [BitSet](https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html)

The second one could save some memory, I guess.

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/SupportsPartitionManagementSuite.scala:
```diff
@@ -140,4 +140,45 @@ class SupportsPartitionManagementSuite extends SparkFunSuite {
+      (Array("part0", "part1"), InternalRow(0, "abc")) -> Set(InternalRow(0, "abc")),
+      (Array("part0"), InternalRow(0)) -> Set(InternalRow(0, "abc"), InternalRow(0, "def")),
+      (Array("part1"), InternalRow("abc")) -> Set(InternalRow(0, "abc"), InternalRow(1, "abc")),
+      (Array.empty[String], InternalRow.empty) ->
```
This is a special case which allows to list all partitions.

cloud-fan commented:
Since `SupportsPartitionManagement` already have the API `partitionSchema`, which means that the implementations will pick a name for partition transforms, I think it's OK to use `String[]` in the `listPartitionIdentifiers` API parameter.

MaxGekk commented:
@cloud-fan @HyukjinKwon Can we continue with this PR or do you have some objections?

cloud-fan commented:
I'm merging it to unblock the following work. @stczwd  @rdblue please leave comments if you have any concerns, so that we can address them.

cloud-fan commented:
merging to master, thanks!

stczwd commented:
> Since `SupportsPartitionManagement` already have the API `partitionSchema`, which means that the implementations will pick a name for partition transforms, I think it's OK to use `String[]` in the `listPartitionIdentifiers` API parameter.

Yeah, I prefer extend the `listPartitionIdentifiers` instead of add new API `listPartitionsByNames`.

rdblue commented:
> The partition can be a transform like year(ts_col), shall we just partition index in the API instead?

If I remember correctly, there should be a schema exposed by the table that describes these. We should get the name from that schema.

MaxGekk commented:
I removed `listPartitionIdentifiers()` and renamed `listPartitionsByNames()` to `listPartitionIdentifiers()`: https://github.com/apache/spark/pull/30514
